grype/db/v6/severity.go

```diff
@@ -0,0 +1,187 @@
+		if err != nil {
+			return vulnerability.UnknownSeverity, fmt.Errorf("unable to parse CVSS vector: %w", err)
+		}
+		if metrics == nil {
+			return vulnerability.UnknownSeverity, nil
+		}
+		return interpretCVSS(metrics.BaseScore, sev.Version), nil
+	default:
+		return vulnerability.UnknownSeverity, nil
+	}
+}
+
+func parseCVSS(vector string) (*vulnerability.CvssMetrics, error) {
+	switch {
+	case strings.HasPrefix(vector, "CVSS:3.0"):
+		cvss, err := gocvss30.ParseVector(vector)
+		if err != nil {
+			return nil, fmt.Errorf("unable to parse CVSS v3 vector: %w", err)
+		}
+		ex := roundScore(cvss.Exploitability())
+		im := roundScore(cvss.Impact())
+		return &vulnerability.CvssMetrics{
+			BaseScore:           roundScore(cvss.BaseScore()),
+			ExploitabilityScore: &ex,
+			ImpactScore:         &im,
+		}, nil
+	case strings.HasPrefix(vector, "CVSS:3.1"):
+		cvss, err := gocvss31.ParseVector(vector)
+		if err != nil {
+			return nil, fmt.Errorf("unable to parse CVSS v3.1 vector: %w", err)
+		}
+		ex := roundScore(cvss.Exploitability())
+		im := roundScore(cvss.Impact())
+		return &vulnerability.CvssMetrics{
+			BaseScore:           roundScore(cvss.BaseScore()),
+			ExploitabilityScore: &ex,
+			ImpactScore:         &im,
+		}, nil
+	case strings.HasPrefix(vector, "CVSS:4.0"):
+		cvss, err := gocvss40.ParseVector(vector)
+		if err != nil {
+			return nil, fmt.Errorf("unable to parse CVSS v4.0 vector: %w", err)
+		}
+		// there are no exploitability and impact scores in CVSS v4.0
+		return &vulnerability.CvssMetrics{
+			BaseScore: roundScore(cvss.Score()),
+		}, nil
+	default:
+		// should be CVSS v2.0 or is invalid
+		cvss, err := gocvss20.ParseVector(vector)
+		if err != nil {
+			return nil, fmt.Errorf("unable to parse CVSS v2 vector: %w", err)
+		}
+		ex := roundScore(cvss.Exploitability())
+		im := roundScore(cvss.Impact())
+		return &vulnerability.CvssMetrics{
+			BaseScore:           roundScore(cvss.BaseScore()),
+			ExploitabilityScore: &ex,
+			ImpactScore:         &im,
+		}, nil
+	}
+}
+
+func roundScore(score float64) float64 {
+	// 3.8870427750000003 -> 3.9
+	return math.Round(score*10) / 10
+}
+
+func interpretCVSS(score float64, version string) vulnerability.Severity {
+	switch version {
+	case "2.0":
+		return interpretCVSSv2(score)
+	case "3.0", "3.1", "4.0":
+		return interpretCVSSv3Plus(score)
+	default:
+		return vulnerability.UnknownSeverity
+	}
+}
+
+func interpretCVSSv2(score float64) vulnerability.Severity {
+	if score == 0 {
+		return vulnerability.NegligibleSeverity
+	}
+	if score > 0 && score <= 3.9 {
+		return vulnerability.LowSeverity
+	}
+	if score >= 4.0 && score <= 6.9 {
+		return vulnerability.MediumSeverity
+	}
+	if score >= 7.0 && score <= 10.0 {
+		return vulnerability.HighSeverity
+	}
+	return vulnerability.UnknownSeverity
+}
+
+func interpretCVSSv3Plus(score float64) vulnerability.Severity {
+	if score == 0 {
+		return vulnerability.NegligibleSeverity
+	}
+	if score > 0 && score <= 3.9 {
```

Comment:
why have the lower bounds on these? I suspect it all works fine after rounding, but it seems easier/safer to just have checks: `if <= 0, if < 4, if < 7, if < 9, if > 10, else`... I was always told floating point comparisons are very finicky due to internal representations that may not be "exact", in the C++ world at least

Reply:
agreed, I'll update. I chose the values based on the official table https://nvd.nist.gov/vuln-metrics/cvss but feels odd.